elastic/kibana
# [Snapshot + Restore] Set snapshots response size limit

This PR sets a limit (`1000`) on the number of snapshots displayed in the snapshots table in Snapshot + Restore and displays a warning if that limit is reached. This is a temporary solution to support scenarios where a user has a large number of snapshots that could potentially create performance issues on the client. In the near future, this will be replaced in favor of server-side pagination.

This PR is supported on the ES side by: https://github.com/elastic/elasticsearch/pull/74451, https://github.com/elastic/elasticsearch/pull/73952.

Also fixes https://github.com/elastic/kibana/issues/44817

### How to test
1. I'm not sure if the ES changes are available via snapshot yet, so you may need to build from source: `yarn es source -E path.repo=/tmp/es_backups`.
1. Create a local FS repository. You will need to define the `max_number_of_snapshots` setting (default is `500`). This setting isn't available in the UI, so it's easiest to set up via Console:

```
PUT /_snapshot/my_repo
{
  "type": "fs",
  "settings": {
    "location": "/tmp/es_backups",
    "max_number_of_snapshots": 1200
  }
}
```

2. Create >1000 snapshots. I've created [a gist](https://gist.github.com/alisonelizabeth/0e570be53764e03dec8c58b7c4e7278f) with instructions to create snapshots in bulk. **This will take some time to run.**
3. Visit the "Snapshots" tab in Snapshot + Restore. Verify the snapshots are still displayed as expected and warning displays.

**Note on ES API changes**
As part of this work, I've replaced the batched snapshot requests in favor of fetching all snapshots in a single request. The ES snapshots API should also now be compatible again on 8.0 and 7.x, and we will no longer need to maintain two different codebases 🎉 (original PR where divergence was introduced: https://github.com/elastic/kibana/pull/39533).

### Screenshots
<img width="2260" alt="Screen Shot 2021-06-28 at 2 16 38 PM" src="https://user-images.githubusercontent.com/5226211/123685609-e2c24b00-d81c-11eb-8310-55834a91f90e.png">

<img width="1538" alt="Screen Shot 2021-06-28 at 2 20 42 PM" src="https://user-images.githubusercontent.com/5226211/123685615-e48c0e80-d81c-11eb-8439-36faf86d89f3.png">

<img width="1429" alt="Screen Shot 2021-06-28 at 8 06 58 PM" src="https://user-images.githubusercontent.com/5226211/123719753-5467bc80-d850-11eb-80ac-fb57d8be32cb.png">

<img width="1458" alt="Screen Shot 2021-06-28 at 8 28 01 PM" src="https://user-images.githubusercontent.com/5226211/123719755-56318000-d850-11eb-827d-b60acad9f56d.png">
